### ISSUE TITLE
GPU Buffer sharing with Godot Shaders

### ISSUE THREAD

**Issue by Ivorforce** [iree-gd/iree.gd#89]
### Describe the proposed feature and its relevance to inferencing

As i understand, there is currently no way to link the inputs or outputs of IREE gpu tasks to Godot shaders (compute, or regular vert / frag). 

### Context or use case

Adding ability to use IREE buffers in godot shaders and vice versa would unlock a whole suit of new capabilities. For one, IREE.gd could be used instead of compute shaders for scientific tasks (i.e. non machine learning tasks).

### Proposed solution

It should be possible to use GPU IREE output tensors / buffers as godot GPU buffers / textures.
It should also be possible to use godot GPU buffers / textures as IREE input tensors / buffers.

**Comment by fire** [comment]
I want this. No idea how to do it.

I assume it'll be like sharing buffers in metal or vulkan.

**Comment by RechieKho** [comment]
IREE tensor is different than Godot's GPU buffer (texture) as IREE tensor could encapsulate multiple dimension of data while texture is width x height x channel count. If one one constricted with image texture then it will definitely be too limiting and not use the full power of IREE. 

It is still possible to convert them manually in Godot. Though if you want to have absolutely native support, well, it is still possible to extend texture class and put GPU computation inside. It won't escape the fate of preprocessing because a lot of tflite model only accept certain dimensions only

**Comment by fire** [comment]
> extend texture class and put IREE GPU computation inside 

@Ivorforce thoughts on this?

**Comment by fire** [comment]
My understanding is we need a ring buffer that syncs the two rates of the iree.gd and the frame sync.

**Comment by RechieKho** [comment]
By manual I mean convert it in CPU space and put it into Godot's shader. It seems like you want to achieve something like both Godot and IREE accessing the same GPU buffer. This is beyond my understanding as well...

**Comment by Ivorforce** [comment]
Yes, for most features I'd use it would be essential to avoid downloading the data to the CPU. You got it, it would mean both would need have to access the same buffer.

Let's keep this issue open for when anybody has a chance to take a look and dissect our options in detail.

**Comment by fire** [comment]
https://github.com/godotengine/godot/pull/98788 @Ivorforce We've plumbing the mesh support in godot engine 4.4 Thoughts?

**Comment by fire** [comment]
@Ivorforce From my point of view the multimesh buffer is a packed byte array, we can reuse the same strategy to share a different packed byte array.

Would that be sufficient? Or do we also need to support tensors with explicit types and shapes?
